Main.hs

```diff
@@ -11,28 +11,47 @@ import qualified Data.List as L
 import Text.Printf (printf)
 
 multByIndex :: [Int] -> [Int]
-multByIndex = undefined
+multByIndex xs = zipWith (*) xs [0..]
 
 powerByIndex :: [Int] -> [Int]
-powerByIndex = undefined
+powerByIndex xs = zipWith (^) xs [0..]
 
+prodList [] = 1
+prodList (h:t) = h * prodList t
+
+short :: [a] -> Bool
+short [] = True
+short [_] = True
+short _ = False
 productOfDifference :: [Int] -> Int
-productOfDifference = undefined
+productOfDifference xs
+    | short xs = error "The list is too short"
+    | otherwise = prodList (zipWith (-) xs (tail xs))
 
-isSorted :: [Int] -> Bool 
-isSorted xs = undefined
+isSorted :: [Int] -> Bool
+isSorted xs = and (zipWith (<=) xs (tail xs))
 
-countElement :: Int -> [Int] -> Int 
-countElement = undefined
+countElement :: Int -> [Int] -> Int
+countElement x [] = 0
+countElement x (y:ys)
+    | x == y = 1 + compareNextNumber
+    | otherwise = compareNextNumber
+    where compareNextNumber = countElement x ys
 
-dotProduct :: [Int] -> [Int] -> Int 
-dotProduct = undefined
+sumList [] = 0
+sumList (h:t) = h + sumList t
+dotProduct :: [Int] -> [Int] -> Int
+dotProduct xs ys = sumList (zipWith (*) xs ys)
 
 applyAll :: [a -> b] -> a -> [b]
-applyAll = undefined
+applyAll [] x = []
+applyAll [f] x = [f x]
+applyAll (f:fs) x = f x : applyAll fs x
 
-interleave :: [a] -> [a] -> [a] 
-interleave = undefined
+interleave :: [a] -> [a] -> [a]
+interleave (x:xs) (y:ys) = x : y : interleave xs ys
+interleave xs [] = []
+interleave [] ys = []
```

Comment:
you can implement this in one line using higher-order functions discussed in the lecture